src/main/java/graphql/execution/DataFetcherExceptionHandler.java

```diff
@@ -1,22 +1,45 @@
 package graphql.execution;
 
+import graphql.ExecutionResult;
 import graphql.PublicSpi;
 import graphql.schema.DataFetcher;
 import graphql.schema.DataFetchingEnvironment;
 
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
 /**
  * This is called when an exception is thrown during {@link graphql.schema.DataFetcher#get(DataFetchingEnvironment)} execution
  */
 @PublicSpi
 public interface DataFetcherExceptionHandler {
 
     /**
-     * When an exception during a call to a {@link DataFetcher} then this handler
-     * is called back to shape the error that should be placed in the list of errors
+     * When an exception occurs during a call to a {@link DataFetcher} then this handler
+     * is called to shape the errors that should be placed in the {@link ExecutionResult#getErrors()}
+     * list of errors.
      *
      * @param handlerParameters the parameters to this callback
      *
      * @return a result that can contain custom formatted {@link graphql.GraphQLError}s
+     *
+     * @deprecated use {@link #handleException(DataFetcherExceptionHandlerParameters)} instead which as an asynchronous
+     * version
      */
+    @Deprecated
     DataFetcherExceptionHandlerResult onException(DataFetcherExceptionHandlerParameters handlerParameters);
+
+    /**
+     * When an exception occurs during a call to a {@link DataFetcher} then this handler
+     * is called to shape the errors that should be placed in the {@link ExecutionResult#getErrors()}
+     * list of errors.
+     *
+     * @param handlerParameters the parameters to this callback
+     *
+     * @return a result that can contain custom formatted {@link graphql.GraphQLError}s
+     */
+    default CompletionStage<DataFetcherExceptionHandlerResult> handleException(DataFetcherExceptionHandlerParameters handlerParameters) {
+        DataFetcherExceptionHandlerResult result = onException(handlerParameters);
+        return CompletableFuture.completedFuture(result);
+    }
```

Comment:
The  consuming code only calls this method.  It delegates back to the old sync method.  So any current implementations will continue to work.

Later at some point we can remove the deprecated version for a breaking change